File: codeforces/1416/D.java
import java.io.BufferedOutputStream;
import java.io.Closeable;
import java.io.DataInputStream;
import java.io.Flushable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.BitSet;
import java.util.InputMismatchException;
import java.util.Objects;
import java.util.function.BiFunction;

/**
 * Built using CHelper plug-in
 * Actual solution is at the top
 */
public class Main {
	static class TaskAdapter implements Runnable {
		@Override
		public void run() {
			long startTime = System.currentTimeMillis();
			InputStream inputStream = System.in;
			OutputStream outputStream = System.out;
			FastReader in = new FastReader(inputStream);
			Output out = new Output(outputStream);
			DGraphAndQueries solver = new DGraphAndQueries();
			solver.solve(1, in, out);
			out.close();
			System.err.println(System.currentTimeMillis()-startTime+"ms");
		}
	}

	public static void main(String[] args) throws Exception {
		Thread thread = new Thread(null, new TaskAdapter(), "", 1<<28);
		thread.start();
		thread.join();
	}

	static class DGraphAndQueries {
		int[] arr;
		int[] parent;
		int[] tin;
		int[] tout;
		int n;
		int ind;
		int clock;
		BitSet visited;
		Pair<Integer, Integer>[] tour;
		ArrayList<Integer>[] graph;

		public DGraphAndQueries() {
		}

		public int find(int u) {
			return parent[u]==u ? u : (parent[u] = find(parent[u]));
		}

		public void merge(Pair<Integer, Integer> p) {
			int u = find(p.a), v = find(p.b);
			if(u==v) {
				return;
			}
			int next = ind++;
			parent[u] = parent[v] = next;
			graph[next] = new ArrayList<>(Arrays.asList(u, v));
		}

		public void dfs(int u) {
			tour[clock] = new Pair<>(arr[u], tin[u] = clock++);
			if(u>=n) {
				for(int v: graph[u]) {
					dfs(v);
				}
			}
			tout[u] = clock-1;
		}

		public void solve(int kase, InputReader in, Output pw) {
			n = in.nextInt();
			int m = in.nextInt(), q = in.nextInt();
			arr = new int[n+q];
			Arrays.fill(arr, -1);
			parent = new int[n+q];
			graph = new ArrayList[n+q];
			ind = n;
			for(int i = 0; i<n; i++) {
				arr[i] = in.nextInt();
			}
			for(int i = 0; i<n+q; i++) {
				parent[i] = i;
			}
			Pair<Integer, Integer>[] edges = new Pair[m];
			for(int i = 0; i<m; i++) {
				edges[i] = new Pair<>(in.nextInt()-1, in.nextInt()-1);
			}
			BitSet marked = new BitSet(m);
			int[][] queries = new int[q][2];
			for(int i = 0; i<q; i++) {
				queries[i] = new int[] {in.nextInt(), in.nextInt()-1};
				if(queries[i][0]==2) {
					marked.set(queries[i][1]);
				}
			}
			for(int i = 0; i<m; i++) {
				if(!marked.get(i)) {
					merge(edges[i]);
				}
			}
			for(int i = q-1; i>=0; i--) {
				if(queries[i][0]==2) {
					merge(edges[queries[i][1]]);
				}else {
					queries[i][1] = find(queries[i][1]);
				}
			}
			Utilities.Debug.dbg(ind);
			Utilities.Debug.dbg(graph);
			tour = new Pair[ind];
			tin = new int[ind];
			tout = new int[ind];
			visited = new BitSet(ind);
			clock = 0;
			for(int i = 0; i<n; i++) {
				int u = find(i);
				if(!visited.get(u)) {
					dfs(u);
					visited.set(u);
				}
			}
			Utilities.Debug.dbg(tour);
			TPointSegmentTree<Pair<Integer, Integer>> st = new TPointSegmentTree<>(tour, (o1, o2) -> o1.a>=o2.a ? o1 : o2);
			for(int[] query: queries) {
				if(query[0]==1) {
					int u = query[1];
					var v = st.query(tin[u], tout[u]);
					Utilities.Debug.dbg(u, v);
					pw.println(v.a);
					st.set(v.b, new Pair<>(0, v.b));
				}
			}
		}

	}

	static class TPointSegmentTree<T> {
		public int n;
		public int ind;
		public int ql;
		public int qr;
		public T[] arr;
		public T[] value;
		public BiFunction<T, T, T> operation;

		public TPointSegmentTree(T[] arr, BiFunction<T, T, T> operation) {
			n = arr.length;
			this.arr = arr;
			value = (T[]) new Object[n<<2];
			this.operation = operation;
			build(1, 0, n-1);
		}

		private void build(int o, int l, int r) {
			if(l==r) {
				value[o] = arr[l];
				return;
			}
			int lc = o<<1, rc = lc|1, mid = l+r >> 1;
			build(lc, l, mid);
			build(rc, mid+1, r);
			value[o] = operation.apply(value[lc], value[rc]);
		}

		private void set(int o, int l, int r) {
			if(l==r) {
				value[o] = arr[l];
				return;
			}
			int lc = o<<1, rc = lc|1, mid = l+r >> 1;
			if(ind<=mid) {
				set(lc, l, mid);
			}else {
				set(rc, mid+1, r);
			}
			value[o] = operation.apply(value[lc], value[rc]);
		}

		public void set(int ind, T val) {
			this.ind = ind;
			arr[ind] = val;
			set(1, 0, n-1);
		}

		private T query(int o, int l, int r) {
			if(ql<=l&&qr>=r) {
				return value[o];
			}
			int lc = o<<1, rc = lc|1, mid = l+r >> 1;
			T ret = null;
			if(ql<=mid) {
				ret = query(lc, l, mid);
			}
			if(qr>mid) {
				if(ret==null) {
					ret = query(rc, mid+1, r);
				}else {
					ret = operation.apply(ret, query(rc, mid+1, r));
				}
			}
			return ret;
		}

		public T query(int l, int r) {
			ql = l;
			qr = r;
			return query(1, 0, n-1);
		}

	}

	static class Utilities {
		public static class Debug {
			public static final boolean LOCAL = System.getProperty("ONLINE_JUDGE")==null;

			private static <T> String ts(T t) {
				if(t==null) {
					return "null";
				}
				try {
					return ts((Iterable) t);
				}catch(ClassCastException e) {
					if(t instanceof int[]) {
						String s = Arrays.toString((int[]) t);
						return "{"+s.substring(1, s.length()-1)+"}";
					}else if(t instanceof long[]) {
						String s = Arrays.toString((long[]) t);
						return "{"+s.substring(1, s.length()-1)+"}";
					}else if(t instanceof char[]) {
						String s = Arrays.toString((char[]) t);
						return "{"+s.substring(1, s.length()-1)+"}";
					}else if(t instanceof double[]) {
						String s = Arrays.toString((double[]) t);
						return "{"+s.substring(1, s.length()-1)+"}";
					}else if(t instanceof boolean[]) {
						String s = Arrays.toString((boolean[]) t);
						return "{"+s.substring(1, s.length()-1)+"}";
					}
					try {
						return ts((Object[]) t);
					}catch(ClassCastException e1) {
						return t.toString();
					}
				}
			}

			private static <T> String ts(T[] arr) {
				StringBuilder ret = new StringBuilder();
				ret.append("{");
				boolean first = true;
				for(T t: arr) {
					if(!first) {
						ret.append(", ");
					}
					first = false;
					ret.append(ts(t));
				}
				ret.append("}");
				return ret.toString();
			}

			private static <T> String ts(Iterable<T> iter) {
				StringBuilder ret = new StringBuilder();
				ret.append("{");
				boolean first = true;
				for(T t: iter) {
					if(!first) {
						ret.append(", ");
					}
					first = false;
					ret.append(ts(t));
				}
				ret.append("}");
				return ret.toString();
			}

			public static void dbg(Object... o) {
				if(LOCAL) {
					System.err.print("Line #"+Thread.currentThread().getStackTrace()[2].getLineNumber()+": [");
					for(int i = 0; i<o.length; i++) {
						if(i!=0) {
							System.err.print(", ");
						}
						System.err.print(ts(o[i]));
					}
					System.err.println("]");
				}
			}

		}

	}

	static class Output implements Closeable, Flushable {
		public StringBuilder sb;
		public OutputStream os;
		public int BUFFER_SIZE;
		public String lineSeparator;

		public Output(OutputStream os) {
			this(os, 1<<16);
		}

		public Output(OutputStream os, int bs) {
			BUFFER_SIZE = bs;
			sb = new StringBuilder(BUFFER_SIZE);
			this.os = new BufferedOutputStream(os, 1<<17);
			lineSeparator = System.lineSeparator();
		}

		public void println(int i) {
			println(String.valueOf(i));
		}

		public void println(String s) {
			sb.append(s);
			println();
		}

		public void println() {
			sb.append(lineSeparator);
		}

		private void flushToBuffer() {
			try {
				os.write(sb.toString().getBytes());
			}catch(IOException e) {
				e.printStackTrace();
			}
			sb = new StringBuilder(BUFFER_SIZE);
		}

		public void flush() {
			try {
				flushToBuffer();
				os.flush();
			}catch(IOException e) {
				e.printStackTrace();
			}
		}

		public void close() {
			flush();
			try {
				os.close();
			}catch(IOException e) {
				e.printStackTrace();
			}
		}

	}

	static interface InputReader {
		int nextInt();

	}

	static class Pair<T1, T2> implements Comparable<Pair<T1, T2>> {
		public T1 a;
		public T2 b;

		public Pair(Pair<T1, T2> p) {
			this(p.a, p.b);
		}

		public Pair(T1 a, T2 b) {
			this.a = a;
			this.b = b;
		}

		public String toString() {
			return a+" "+b;
		}

		public int hashCode() {
			return Objects.hash(a, b);
		}

		public boolean equals(Object o) {
			if(o instanceof Pair) {
				Pair p = (Pair) o;
				return a.equals(p.a)&&b.equals(p.b);
			}
			return false;
		}

		public int compareTo(Pair<T1, T2> p) {
			int cmp = ((Comparable<T1>) a).compareTo(p.a);
			if(cmp==0) {
				return ((Comparable<T2>) b).compareTo(p.b);
			}
			return cmp;
		}

	}

	static class FastReader implements InputReader {
		final private int BUFFER_SIZE = 1<<16;
		private DataInputStream din;
		private byte[] buffer;
		private int bufferPointer;
		private int bytesRead;

		public FastReader(InputStream is) {
			din = new DataInputStream(is);
			buffer = new byte[BUFFER_SIZE];
			bufferPointer = bytesRead = 0;
		}

		public int nextInt() {
			int ret = 0;
			byte c = skipToDigit();
			boolean neg = (c=='-');
			if(neg) {
				c = read();
			}
			do {
				ret = ret*10+c-'0';
			} while((c = read())>='0'&&c<='9');
			if(neg) {
				return -ret;
			}
			return ret;
		}

		private boolean isDigit(byte b) {
			return b>='0'&&b<='9';
		}

		private byte skipToDigit() {
			byte ret;
			while(!isDigit(ret = read())&&ret!='-') ;
			return ret;
		}

		private void fillBuffer() {
			try {
				bytesRead = din.read(buffer, bufferPointer = 0, BUFFER_SIZE);
			}catch(IOException e) {
				e.printStackTrace();
				throw new InputMismatchException();
			}
			if(bytesRead==-1) {
				buffer[0] = -1;
			}
		}

		private byte read() {
			if(bytesRead==-1) {
				throw new InputMismatchException();
			}else if(bufferPointer==bytesRead) {
				fillBuffer();
			}
			return buffer[bufferPointer++];
		}

	}
}

File: codeforces/1416/D.kt
import java.io.*
import java.util.*
import java.util.function.BiFunction
import kotlin.collections.ArrayList

/**
 * Built using CHelper plug-in
 * Actual solution is at the top
 */
fun main(args: Array<String>) {
    val thread = Thread(null, TaskAdapter(), "", 1 shl 28)
    thread.start()
    thread.join()
}

internal class TaskAdapter : Runnable {
    override fun run() {
        val startTime = System.currentTimeMillis()
        val inputStream = System.`in`
        val outputStream: OutputStream = System.out
        val `in` = FastReader(inputStream)
        val out = Output(outputStream)
        val solver = DGraphAndQueries()
        solver.solve(1, `in`, out)
        out.close()
        System.err.println((System.currentTimeMillis() - startTime).toString() + "ms")
    }
}

internal class DGraphAndQueries {
    lateinit var arr: IntArray
    lateinit var parent: IntArray
    lateinit var tin: IntArray
    lateinit var tout: IntArray
    var n = 0
    var ind = 0
    var clock = 0
    lateinit var visited: BitSet
    lateinit var tour: Array<Pair<Int, Int>>
    lateinit var graph: Array<ArrayList<Int>>
    fun find(u: Int): Int {
        return if (parent[u] == u) u else find(parent[u]).also { parent[u] = it }
    }

    fun merge(p: Pair<Int, Int>) {
        val u = find(p.a)
        val v = find(p.b)
        if (u == v) {
            return
        }
        val next = ind++
        parent[v] = next
        parent[u] = parent[v]
        graph[next] = ArrayList(Arrays.asList(u, v))
    }

    fun dfs(u: Int) {
        tour[clock] = Pair(arr[u], clock++.also { tin[u] = it })
        if (u >= n) {
            for (v in graph[u]) {
                dfs(v)
            }
        }
        tout[u] = clock - 1
    }

    fun solve(kase: Int, `in`: InputReader, pw: Output) {
        n = `in`.nextInt()
        val m = `in`.nextInt()
        val q = `in`.nextInt()
        arr = IntArray(n + q)
        Arrays.fill(arr, -1)
        parent = IntArray(n + q)
        graph = Array(n + q) { ArrayList<Int>() }
        ind = n
        for (i in 0 until n) {
            arr[i] = `in`.nextInt()
        }
        for (i in 0 until n + q) {
            parent[i] = i
        }
        val edges: Array<Pair<Int, Int>> = Array(m) { Pair(0, 0) }
        for (i in 0 until m) {
            edges[i] = Pair(`in`.nextInt() - 1, `in`.nextInt() - 1)
        }
        val marked = BitSet(m)
        val queries = Array(q) { IntArray(2) }
        for (i in 0 until q) {
            queries[i] = intArrayOf(`in`.nextInt(), `in`.nextInt() - 1)
            if (queries[i][0] == 2) {
                marked.set(queries[i][1])
            }
        }
        for (i in 0 until m) {
            if (!marked[i]) {
                merge(edges[i])
            }
        }
        for (i in q - 1 downTo 0) {
            if (queries[i][0] == 2) {
                merge(edges[queries[i][1]])
            } else {
                queries[i][1] = find(queries[i][1])
            }
        }
//            Utilities.Debug.dbg(ind)
//            Utilities.Debug.dbg(*graph)
        tour = Array(ind) { Pair(0, 0) }
        tin = IntArray(ind)
        tout = IntArray(ind)
        visited = BitSet(ind)
        clock = 0
        for (i in 0 until n) {
            val u = find(i)
            if (!visited[u]) {
                dfs(u)
                visited.set(u)
            }
        }
//            Utilities.Debug.dbg(*tour)
        val st = TPointSegmentTree(tour, BiFunction { t, u -> if (t.a >= u.a) t else u })
        for (query in queries) {
            if (query[0] == 1) {
                val u = query[1]
                val v = st.query(tin[u], tout[u])!!
                Utilities.Debug.dbg(u, v)
                pw.println(v.a)
                st[v.b] = Pair(0, v.b)
            }
        }
    }
}

internal class TPointSegmentTree<T>(arr: Array<T>, operation: BiFunction<T, T, T>) {
    var n: Int
    var ind = 0
    var ql = 0
    var qr = 0
    var arr: Array<T>
    var value: Array<T?>
    var operation: BiFunction<T, T, T>
    private fun build(o: Int, l: Int, r: Int) {
        if (l == r) {
            value[o] = arr[l]
            return
        }
        val lc = o shl 1
        val rc = lc or 1
        val mid = l + r shr 1
        build(lc, l, mid)
        build(rc, mid + 1, r)
        value[o] = value[lc]?.let { value[rc]?.let { it1 -> operation.apply(it, it1) } }
    }

    private operator fun set(o: Int, l: Int, r: Int) {
        if (l == r) {
            value[o] = arr[l]
            return
        }
        val lc = o shl 1
        val rc = lc or 1
        val mid = l + r shr 1
        if (ind <= mid) {
            set(lc, l, mid)
        } else {
            set(rc, mid + 1, r)
        }
        value[o] = value[lc]?.let { value[rc]?.let { it1 -> operation.apply(it, it1) } }
    }

    operator fun set(ind: Int, `val`: T) {
        this.ind = ind
        arr[ind] = `val`
        set(1, 0, n - 1)
    }

    private fun query(o: Int, l: Int, r: Int): T? {
        if (ql <= l && qr >= r) {
            return value[o]
        }
        val lc = o shl 1
        val rc = lc or 1
        val mid = l + r shr 1
        var ret: T? = null
        if (ql <= mid) {
            ret = query(lc, l, mid)
        }
        if (qr > mid) {
            ret = if (ret == null) {
                query(rc, mid + 1, r)
            } else {
                query(rc, mid + 1, r)?.let { operation.apply(ret!!, it) }
            }
        }
        return ret
    }

    fun query(l: Int, r: Int): T? {
        ql = l
        qr = r
        return query(1, 0, n - 1)
    }

    init {
        n = arr.size
        this.arr = arr
        value = arrayOfNulls<Any>(n shl 2) as Array<T?>
        this.operation = operation
        build(1, 0, n - 1)
    }
}

internal class Utilities {
    object Debug {
        val LOCAL = System.getProperty("ONLINE_JUDGE") == null
        private fun <T> ts(t: T?): String {
            return if (t == null) {
                "null"
            } else try {
                ts<Any>(t as Iterable<*>)
            } catch (e: ClassCastException) {
                if (t is IntArray) {
                    val s = Arrays.toString(t as IntArray?)
                    return "{" + s.substring(1, s.length - 1) + "}"
                } else if (t is LongArray) {
                    val s = Arrays.toString(t as LongArray?)
                    return "{" + s.substring(1, s.length - 1) + "}"
                } else if (t is CharArray) {
                    val s = Arrays.toString(t as CharArray?)
                    return "{" + s.substring(1, s.length - 1) + "}"
                } else if (t is DoubleArray) {
                    val s = Arrays.toString(t as DoubleArray?)
                    return "{" + s.substring(1, s.length - 1) + "}"
                } else if (t is BooleanArray) {
                    val s = Arrays.toString(t as BooleanArray?)
                    return "{" + s.substring(1, s.length - 1) + "}"
                }
                try {
                    ts(t as Array<Any>)
                } catch (e1: ClassCastException) {
                    t.toString()
                }
            }
        }

        private fun <T> ts(arr: Array<T>): String {
            val ret = StringBuilder()
            ret.append("{")
            var first = true
            for (t in arr) {
                if (!first) {
                    ret.append(", ")
                }
                first = false
                ret.append(ts(t))
            }
            ret.append("}")
            return ret.toString()
        }

        private fun <T> ts(iter: Iterable<T>): String {
            val ret = StringBuilder()
            ret.append("{")
            var first = true
            for (t in iter) {
                if (!first) {
                    ret.append(", ")
                }
                first = false
                ret.append(ts(t))
            }
            ret.append("}")
            return ret.toString()
        }

        fun dbg(vararg o: Any?) {
            if (LOCAL) {
                System.err.print("Line #" + Thread.currentThread().stackTrace[2].lineNumber + ": [")
                for (i in 0 until o.size) {
                    if (i != 0) {
                        System.err.print(", ")
                    }
                    System.err.print(ts<Any?>(o[i]))
                }
                System.err.println("]")
            }
        }
    }
}

internal class Output @JvmOverloads constructor(os: OutputStream?, var BUFFER_SIZE: Int = 1 shl 16) : Closeable, Flushable {
    var sb: StringBuilder
    var os: OutputStream
    var lineSeparator: String
    fun println(i: Int) {
        println(i.toString())
    }

    fun println(s: String?) {
        sb.append(s)
        println()
    }

    fun println() {
        sb.append(lineSeparator)
    }

    private fun flushToBuffer() {
        try {
            os.write(sb.toString().toByteArray())
        } catch (e: IOException) {
            e.printStackTrace()
        }
        sb = StringBuilder(BUFFER_SIZE)
    }

    override fun flush() {
        try {
            flushToBuffer()
            os.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun close() {
        flush()
        try {
            os.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    init {
        sb = StringBuilder(BUFFER_SIZE)
        this.os = BufferedOutputStream(os, 1 shl 17)
        lineSeparator = System.lineSeparator()
    }
}

internal interface InputReader {
    fun nextInt(): Int
}

internal class Pair<T1, T2>(var a: T1, var b: T2) : Comparable<Pair<T1, T2>> {
    constructor(p: Pair<T1, T2>) : this(p.a, p.b) {}

    override fun toString(): String {
        return a.toString() + " " + b
    }

    override fun hashCode(): Int {
        return Objects.hash(a, b)
    }

    override fun equals(o: Any?): Boolean {
        if (o is Pair<*, *>) {
            val p = o
            return a == p.a && b == p.b
        }
        return false
    }

    override fun compareTo(p: Pair<T1, T2>): Int {
        val cmp = (a as Comparable<T1>).compareTo(p.a)
        return if (cmp == 0) {
            (b as Comparable<T2>).compareTo(p.b)
        } else cmp
    }
}

internal class FastReader(`is`: InputStream?) : InputReader {
    private val BUFFER_SIZE = 1 shl 16
    private val din: DataInputStream
    private val buffer: ByteArray
    private var bufferPointer: Int
    private var bytesRead: Int
    override fun nextInt(): Int {
        var ret = 0
        var c = skipToDigit()
        val neg = c == '-'.toByte()
        if (neg) {
            c = read()
        }
        do {
            ret = ret * 10 + c - '0'.toInt()
        } while (read().also { c = it } >= '0'.toByte() && c <= '9'.toByte())
        return if (neg) {
            -ret
        } else ret
    }

    private fun isDigit(b: Byte): Boolean {
        return b >= '0'.toByte() && b <= '9'.toByte()
    }

    private fun skipToDigit(): Byte {
        var ret: Byte
        while (!isDigit(read().also { ret = it }) && ret != '-'.toByte());
        return ret
    }

    private fun fillBuffer() {
        try {
            bytesRead = din.read(buffer, 0.also { bufferPointer = it }, BUFFER_SIZE)
        } catch (e: IOException) {
            e.printStackTrace()
            throw InputMismatchException()
        }
        if (bytesRead == -1) {
            buffer[0] = -1
        }
    }

    private fun read(): Byte {
        if (bytesRead == -1) {
            throw InputMismatchException()
        } else if (bufferPointer == bytesRead) {
            fillBuffer()
        }
        return buffer[bufferPointer++]
    }

    init {
        din = DataInputStream(`is`)
        buffer = ByteArray(BUFFER_SIZE)
        bytesRead = 0
        bufferPointer = bytesRead
    }
}
